src/Linters/DontShadowLocalVariablesInForeachLinter.hack

```diff
@@ -0,0 +1,164 @@
+
+use namespace HH\Lib\{C, Dict, Keyset, Str, Vec};
+
+final class DontShadowLocalVariablesInForeachLinter extends ASTLinter {
+  const type TContext = IFunctionishDeclaration;
+  const type TNode = ForeachStatement;
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    this::TContext $scope,
+    this::TNode $foreach,
+  ): ?ASTLintError {
+    $pseudo_declarations = static::getPseudoDeclarations($foreach);
+    if (C\is_empty($pseudo_declarations)) {
+      return null;
+    }
+    $scope_decl = i_functionish_get_decl_header($scope);
+    $params = $scope_decl->getParameterList()
+      ?->getDescendantsOfType(VariableToken::class)
+      |> Keyset\map($$ ?? vec[], $p ==> $p->getText());
+
+    unset($params['$_']);
+
+    $first_dup = Dict\select_keys($pseudo_declarations, $params) |> C\first($$);
+    if ($first_dup is nonnull) {
+      return $this->paramDup($scope_decl->getName()->getText(), $first_dup);
+    }
+
+    $lambda_params = $this->getParamsFromEnclosingLambdas($foreach);
+    unset($lambda_params['$_']);
+
+    $first_dup = Dict\select_keys($pseudo_declarations, $lambda_params)
+      |> C\first($$);
+    if ($first_dup is nonnull) {
+      return $this->paramDup('an anonymous function in your scope', $first_dup);
+    }
+
+    $locals = static::getAssignedLocalsBeforeForeach(
+      $scope->getBody(),
+      $foreach->getKeyword(),
+    );
+    unset($locals['$_']);
+
+    foreach ($locals as $var => $expr) {
+      if (C\contains_key($pseudo_declarations, $var)) {
+        return $this->localVariableDup($expr, $pseudo_declarations[$var]);
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * This method is a best effort.
+   * `foreach(vec[] as $dict[$key]) {}` is valid Hack.
+   * We count `$key` as a declaration, but that is incorrect.
+   */
+  private static function getPseudoDeclarations(
+    this::TNode $foreach,
+  ): dict<string, VariableToken> {
+    $pseudo_declarations = dict[];
+
+    $key = $foreach->getKey();
+    if ($key is VariableExpression) {
+      $pseudo_declarations = $key->getDescendantsOfType(VariableToken::class)
+        |> Dict\from_values($$, $token ==> $token->getText());
+    }
+
+    $value = $foreach->getValue();
+    $value_vars = $value->getDescendantsOfType(VariableToken::class)
+      |> Dict\from_values($$, $token ==> $token->getText());
+
+    if (C\is_empty($value_vars)) {
+      return $pseudo_declarations;
+    }
+    return Dict\merge($value_vars, $pseudo_declarations);
+  }
+
+  private function getParamsFromEnclosingLambdas(
+    ForeachStatement $foreach,
+  ): keyset<string> {
+    $lambdas = $this->getAST()
+      ->getAncestorsOfDescendantOfType<LambdaExpression>($foreach);
+
+    return Vec\map(
+      $lambdas,
+      $lambda ==>
+        $lambda->getSignature()->getDescendantsOfType(VariableToken::class),
+    )
+      |> Vec\flatten($$)
+      |> Keyset\map($$, $p ==> $p->getText());
+  }
+
+  private static function getAssignedLocalsBeforeForeach(
```

Comment:
This looks potentially similar to UnusedVariableLinter::classifyVariables, can we extract/reuse it?